### DISCUSSION TITLE
isUtf8(buffer, start?, end?)

### DISCUSSION THREAD

**Discussion by micnic** [nodejs/help#4672] [Ideas]
Would it make sense to add 2 more arguments to [isUtf8()](https://nodejs.org/api/buffer.html#bufferisutf8input)?

```ts
declare function isUtf8(buffer: Uint8Array, start?: number, end?: number): boolean;
```

Where `start` would default to `0` and `end` would default to `buffer.length`. It should validate only the provided range, or full buffer if no arguments are provided. This would avoid function users to create a buffer slice that comes at a cost.

**Comment by marco-ippolito** [comment]
This sounds like a reasonable feature request, I'd be +1

**Comment by aduh95** [comment]
Is there an advantage for doing that rather than `isUtf8(buffer.subarray(start, end))`?

**Reply by micnic** [reply]
Sure, it is much faster to read a buffer in a range than to create a new buffer instance with offset from the original one. I created a benchmark here for an ASCII validation (this approach can be used for `NodeBuffer.isAscii()` too):

https://esbench.com/bench/65a399847ff73700a4debf72

**Reply by targos** [reply]
Your benchmark is testing an extremely small buffer. I expect the speed difference to depend largely on the buffer length (and range check length too).

**Reply by micnic** [reply]
I updated the benchmark with a larger buffer, and received similar results. Not sure we'll get any different outcome if we increase the buffer more.

https://esbench.com/bench/65a399847ff73700a4debf72

**Reply by aduh95** [reply]
OK interesting, as `TypedArray.prototype.subarray` does not copy memory (reuse the same underlying `ArrayBuffer`), I would have expected the speed to be about the same; I guess there's some overhead in creating a new `TypedArray` instance indeed.
